test/e2e/network/netpol/reachability.go

```diff
@@ -0,0 +1,184 @@
+func NewReachability(pods []*Pod, defaultExpectation bool) *Reachability {
+	var podNames []string
+	for _, pod := range pods {
+		podNames = append(podNames, pod.PodString().String())
+	}
+	r := &Reachability{
+		Expected: NewTruthTableFromItems(podNames, &defaultExpectation),
+		Observed: NewTruthTableFromItems(podNames, nil),
+		Pods:     pods,
+	}
+	return r
+}
+
+// AllowLoopback expects all communication from a pod to itself to be allowed.
+// In general, call it after setting up any other rules since loopback logic follows no policy.
+func (r *Reachability) AllowLoopback() {
+	for _, pod := range r.Pods {
+		podName := pod.PodString().String()
+		r.Expected.Set(podName, podName, true)
+	}
+}
+
+// Expect sets the expected value for a single observation
+func (r *Reachability) Expect(from PodString, to PodString, isConnected bool) {
+	r.Expected.Set(string(from), string(to), isConnected)
+}
+
+// ExpectAllIngress defines that any traffic going into the pod will be allowed/denied (true/false)
+func (r *Reachability) ExpectAllIngress(pod PodString, connected bool) {
+	r.Expected.SetAllTo(string(pod), connected)
+	if !connected {
+		log.Infof("Denying all traffic *to* %s", pod)
+	}
+}
+
+// ExpectAllEgress defines that any traffic going out of the pod will be allowed/denied (true/false)
+func (r *Reachability) ExpectAllEgress(pod PodString, connected bool) {
+	r.Expected.SetAllFrom(string(pod), connected)
+	if !connected {
+		log.Infof("Denying all traffic *from* %s", pod)
+	}
+}
+
+// ExpectPeer sets expected values using Peer matchers
+func (r *Reachability) ExpectPeer(from *Peer, to *Peer, connected bool) {
+	for _, fromPod := range r.Pods {
+		if from.Matches(fromPod.PodString()) {
+			for _, toPod := range r.Pods {
+				if to.Matches(toPod.PodString()) {
+					r.Expected.Set(string(fromPod.PodString()), string(toPod.PodString()), connected)
+				}
+			}
+		}
+	}
+}
+
+// Observe records a single connectivity observation
+func (r *Reachability) Observe(fromPod PodString, toPod PodString, isConnected bool) {
+	r.Observed.Set(string(fromPod), string(toPod), isConnected)
+}
+
+// Summary produces a useful summary of expected and observed data
+func (r *Reachability) Summary() (trueObs int, falseObs int, comparison *TruthTable) {
+	comparison = r.Expected.Compare(r.Observed)
+	if !comparison.IsComplete() {
+		panic("observations not complete!")
+	}
+	falseObs = 0
+	trueObs = 0
+	for _, dict := range comparison.Values {
+		for _, val := range dict {
+			if val {
+				trueObs++
+			} else {
+				falseObs++
+			}
+		}
+	}
+	return trueObs, falseObs, comparison
+}
+
+// PrintSummary prints the summary
+func (r *Reachability) PrintSummary(printExpected bool, printObserved bool, printComparison bool) {
+	right, wrong, comparison := r.Summary()
+	fmt.Printf("reachability: correct:%v, incorrect:%v, result=%t\n\n", right, wrong, wrong == 0)
+	if printExpected {
+		fmt.Printf("expected:\n\n%s\n\n\n", r.Expected.PrettyPrint(""))
```

Comment:
printf does not work inside ginkgo xref https://github.com/onsi/ginkgo/issues/429
you can use the framework logging tools

Reply:
👍